fairscale/nn/pipe/messages.py

```diff
@@ -0,0 +1,137 @@
+import pickle
+from queue import Empty as QueueEmpty
+from queue import Queue
+from typing import Any, List
+
+import numpy as np
+import torch
+from torch import Tensor
+
+from fairscale.nn.model_parallel import get_pipeline_parallel_group
+
+from .types import MESSAGE_GENERATION_START, InputDevice, PipeMessage, Tensors, TransportConfig
+
+# FIXME Why is 256 ok for training but not for tests?
+MESSAGE_TENSOR_SIZE = 1024  # 256
+
+MessageQueues: List[Queue] = [Queue() for _ in range(MESSAGE_GENERATION_START)]
+
+
+def pyobject_to_tensor(obj: Any) -> Tensor:
+    pickled = pickle.dumps(obj)
+    nparray = np.frombuffer(pickled, dtype=np.uint8).copy()
+    nparray.setflags(write=True)
+    result = torch.from_numpy(nparray)
+    delta = MESSAGE_TENSOR_SIZE - len(result)
+    if delta < 0:
+        raise ValueError(
+            f"message too big to send, increase MESSAGE_TENSOR_SIZE? - {len(result)} > {MESSAGE_TENSOR_SIZE}"
+        )
+    elif delta > 0:
+        result = torch.cat((result, torch.zeros(delta, dtype=torch.uint8)))
+
+    return result.cuda()
+
+
+def tensor_to_pyobject(tensor: Tensor) -> Any:
+    try:
+        nparray = tensor.numpy()
+        return pickle.loads(nparray.tobytes())
+    except Exception as e:
+        print(f"pickle fail {e}")
+        raise e
+
+
+def to_input_device(tensors: Tensors, input_device: InputDevice) -> Tensors:
+    if input_device is None:
+        return tensors
+    else:
+        return tuple(t.to(input_device) for t in tensors)
+
+
+def rpc_push_queue(message: PipeMessage) -> None:
+    globals()["MessageQueues"][message.queue_name].put(message)
+
+
+def send_message(config: TransportConfig, message: PipeMessage, sync: bool = False) -> None:
```

Comment:
I don't see a test for this or any callers in this PR. Is this dead code or is their a follow-up PR that uses this?

Reply:
It's called from pipeline.py and async_schedule.py